### DISCUSSION TITLE
Joining several devices

### DISCUSSION THREAD

**Discussion by Anders-Holst** [Anders-Holst/xled_plus#11] [Effects]
Many of the recent daily effects assume or work best on dense 2D displays. A typical situation would be when you have a wall covered with leds. However, then you typically have several Twinkly lights connected as a unit, and so far the HighControlInterface used from python assumes a single device.

I have written a class MultiHighControlInterface, and included in the file "multicontrol.py" in xled_plus from today's release (Dec 30), in an attempt to solve this. Assume that you have used the app to join several devices, and produce a joint layout mapping for them by some of the methods offered on the app. Then you can create a combined control object for them as:
```
from xled_plus.multicontrol import MultiHighControlInterface
mctr = MultiHighControlInterface(["ip-address-master", "ip-address-slave1",..., "ip-address-slaveN"])
```
where you insert the ip-addresses of the connected devices, starting with the master. (How do you know which is the master? Probably the device on the first line in the app. It was in my case. Maybe I should fix the code so it finds out itself. I need to get back on this one.) Then you can use `mctr` just as you would have with a normal HighControlInterface object:
```
from xled_plus.samples.sample_setup import *
cols = [hsl_color(0.375, 0.2, 0.0), hsl_color(0.375, 1.0, 0.0), hsl_color(0.28, 1.0, 0.0), hsl_color(0.375, 0.2, 0.0), hsl_color(0.375, 1.0, 0.0), hsl_color(0.5, 1.0, 0.0)]
eff = GradientSequence(mctr, cols, speed=0.04, folds=0.5, angle=45)
eff.launch_movie()
```
This is an experimental code, and I don't have any large wall with lights myself, so I appreciate any feedback and bug reports!

 / Anders

**Comment by Anders-Holst** [comment]
I have now made an attempt to fix the MultiHighControlInterface so it can find the master itself. Thus you can provide the ip-addresses in any order. Untested... please let me know if there are problems.

**Comment by DanDixon** [comment]
I got another set of twinkly lights and just tried to connect them.

The code runs without error and both light strings react, but only the slave seems to actually get the effect. The master returns to the effect it was running previously. If I use a different effect sometimes they both go black.

The MultiHighControlInterface doesn't seem to have all of the functions of the HighControlInterface (or am I missing something). Should the Multi interface be able to do everything the High interface does (or is it unfinished)?

I'm new to python in the last few weeks, so still struggling to make sense of it all. Thanks again for creating and sharing this library. I'm having fun in between all the moments that aren't. 😏

**Comment by DanDixon** [comment]
I spent some more time trying to get this working after adding a 3rd set of lights (this is getting slightly out of hand).

I couldn't get the MultiHighControlInterface to work as the lights just go black and revert to the last setting. I'd grouped them with the Sync method in the Twinkly app. They're all RGB sets. I got a RGBW set, but didn't like how the bulbs looked. The diffuser on those didn't fully cover the led and so there were bright spots that I didn't care for. And a very similar "warm" look (the claimed advantage of the RGBW bulbs) can also be achieved by setting the RGB bulbs to be slightly orange.

My workaround:
I'm calling a shell script from home assistant, so ended up just calling a command for each string. That seems to work well enough and only takes 3-4 seconds longer to set all 3 strings. I added a '&' at the end of each shh line, which I believe calls them in parallel and that seems to be working well.

Cheers!

**Comment by Anders-Holst** [comment]
That version of multicontrol was experimental and not tested so much. I have now checked in a new version (in this repo - but not released it yet since I will do some more changes in other files, so you need to download it from here) that is at least a little more tested. I updated highcontrol.py as well, you will need those fixes too I think. Try it and see if it works better!

Of course it is possible to start the three lights separately in many cases. It may just be more convenient if this multicontrol works as intended. And continuous effects will be difficult without it.

**Comment by DanDixon** [comment]
Thanks for spending time on this.

Unfortunately, I'm getting this error when I try the day1 effect without using any of my modified code:

```
dan@Twinkly:~$ python3 -m xled_plus.samples.day1 192.168.1.23
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/dan/.local/lib/python3.10/site-packages/xled_plus/samples/day1.py", line 14, in <module>
    SparkleStars(ctr).launch_movie()
  File "/home/dan/.local/lib/python3.10/site-packages/xled_plus/effect_base.py", line 95, in launch_movie
    self.stop_rt()
  File "/home/dan/.local/lib/python3.10/site-packages/xled_plus/effect_base.py", line 85, in stop_rt
    self.ctr.set_mode(self.ctr.last_mode)
  File "/home/dan/.local/lib/python3.10/site-packages/xled_plus/highcontrol.py", line 245, in set_mode
    assert mode in ("movie", "playlist", "rt", "demo", "effect", "color", "off")
AssertionError
```

If I comment out line 85, in stop_rt
`    self.ctr.set_mode(self.ctr.last_mode)`
Then it works.

Without commenting out the line, I get the same error as above when I try using a custom glow effect.

The first couple of times I tried my modified code with that line commented out, the first light strand worked, but the other 2 strands didn't.

But then I tried it again (with the above line commented out) and the custom effects worked on all 3 strands. I tried a few more and it seems like it's working great!

Either way, seems like there's an issue with last_mode?

Thanks again so much!

**Comment by Anders-Holst** [comment]
It appears that I managed to add a bug with my modification of effect_base. Thanks for detecting it. I now believe I have fixed it by wrapping line 85 with an if-statement to only run it if last_mode is set. 

(Did you send another message too? I got notified by mail, but it is not here. Maybe you sorted it out yourself already?)

**Comment by DanDixon** [comment]
Great! Those changes work. Thank you!

The other message I posted, I then deleted because I wondered if the problem was related to me commenting out that line of code. However, it doesn't seem to be the case. After sending multiple movies to the strings, I eventually get this:

```
...
  File "/home/dan/.local/lib/python3.10/site-packages/xled/response.py", line 65, in data
    self.response.raise_for_status()
  File "/usr/lib/python3/dist-packages/requests/models.py", line 943, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 406 Client Error: Not Acceptable for url: http://192.168.1.88/xled/v1/movies/new
```

Looks like I'm filling up the memory of the lights?
Interestingly, this is an error on the first secondary device (192.168.1.88), not the primary (which is 192.168.1.23)

If I opened up the Twinkly app > Home (top left) > Settings > Erase Device Effects
Then the error goes away. 

Maybe I just need to call this on occasion? (haven't tried this yet)
'def delete_movies(self):' 

---

As an aside, sending a movie to 3 strings takes 15+ seconds, way slower than sending to an individual string (and way slower than my workaround of calling your code 3 times in parallel).

**Reply by Anders-Holst** [reply]
Aha, I see. Yes, there is a limit on the number of movies that can be uploaded. The limit depends on the model and firmware version. For a while it was 15, in a later version it increased. The python code is supposed to check this limit and call delete_movies when needed (yes you are right, it has the same effect as "Erase Device Effects" in the app). However, I was lazy and assumed all grouped strings to have the same limit, so I only check the master. 

Thus, if the second string has a lower limit, it will cause this error. Is this the case, that the second device is another firmware version or model? (If you have a python prompt with a `HighControlInterface` to each of the two devices, you can check the variable `max_movies` in each.) It would be easy to fix by just using the minimum spec from all devices. I will do that next time I check in. 

Ok, is really that much slower? I am not sure what takes the time then. I had expected it to be pretty much equal time as uploading the three parts in sequence. Actual communication with the devices should take the same time. But python is a really slow language, and I may not have used the fastest tricks everywhere.

**Reply by DanDixon** [reply]
All 3 of my light strings report the same info:

max_movies 55
num_leds 600
family M
led_profile RGB
version (2, 8, 15)

No worries about the speed; I'm just reporting my findings. I'm thrilled that it works. Thanks again for sharing this.

Let me know if there's any other info that would be helpful.
